lte/gateway/python/magma/pipelined/app/apn.py

```diff
@@ -0,0 +1,284 @@
+"""
+Copyright (c) 2016-present, Facebook, Inc.
+All rights reserved.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree. An additional grant
+of patent rights can be found in the PATENTS file in the same directory.
+"""
+
+import logging
+from magma.pipelined.app.base import MagmaController, ControllerType
+from magma.pipelined.openflow import flows
+from magma.pipelined.openflow.registers import (
+    Direction, 
+    APN_TAG_REG, 
+    XXREGISTERS_MAP)
+from magma.pipelined.openflow.magma_match import MagmaMatch
+from magma.pipelined.apn import encode_apn, split_apn
+from magma.pipelined.openflow.messages import MsgChannel, MessageHub
+
+from ryu.controller.controller import  Datapath
+from ryu.ofproto import nicira_ext
+from ryu.ofproto import ether
+from ryu.controller.handler import MAIN_DISPATCHER, set_ev_cls
+from ryu.controller import ofp_event
+
+
+class APNController(MagmaController):
+    """
+    APN Controller.
+
+    The APN controller is responsible for marking a flow with an encoded APN 
+    name. The APN name should be stored in
+    register.
+
+    """
+
+    APP_NAME = "apn"
+    APP_TYPE = ControllerType.LOGICAL
+
+    def __init__(self, *args, **kwargs):
+        super(APNController, self).__init__(*args, **kwargs)
+        # TODO(119vik) get following from the config file
+        self._clean_start = True 
+        self._datapath = None
+        self.tbl_num = self._service_manager.get_table_num(self.APP_NAME)
+        self.apn_tagging_scratch = \
+            self._service_manager.allocate_scratch_tables(self.APP_NAME, 1)[0]
+        self.next_table = self._service_manager.get_next_table_num(
+            self.APP_NAME)
+        self._msg_hub = MessageHub(self.logger)
+
+    def initialize_on_connect(self, datapath):
+        """
+        Install the default flows on datapath connect event.
+
+        Args:
+            datapath: ryu datapath struct
+        """
+        self._datapath = datapath
+        # In case wee need to clean all existing  buggy / orphaned flows 
+        # before start the controller
+        if self._clean_start:
```

Comment:
always nuke all flows on startup, you can just remove the _clean_start var

Reply:
thanks, fixed